### PR TITLE
fixUpJoinConsistency rule now works when AQE is enabled

### PR DESCRIPTION
**Pull request by andygrove**
`GpuOverrides` applies a `fixUpJoinConsistency` rule to ensure that the inputs to a `SortMergeJoin` or `ShuffledHashJoin` are either both on CPU, or both on GPU. We can't support a mix of CPU and GPU because the hashing algorithms are not compatible, and therefore the join would produce incorrect results.

This PR adds a unit test for this rule, and also ensures that the rule is applied when AQE is enabled.

This closes https://github.com/NVIDIA/spark-rapids/issues/631

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 1]
```diff
@@ -471,17 +479,30 @@ abstract class SparkPlanMeta[INPUT <: SparkPlan](plan: INPUT,
       }
     } else if (exchanges.isEmpty) {
       // we only have query stages, and we cannot do anything to modify them at this point
-      // since they already started to execute
+      // since they already started to execute, but we verify that they are both on CPU or
+      // both on GPU
+      if (queryStages.map(isGpuQueryStage).distinct.size == 2) {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:484]: so for my edification, we can't get in this situation unless there is a bug, or perhaps a change in behavior from AQE rules proper?

**Review by abellina — COMMENTED** [review]
Just had one question

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
@abellina I was able to greatly simplify the logic in this PR. Please take a look when you can.

**Pull request merged**